[doc] Move cleanup part from README to documentation (partially merged in #1491) (#1469)
interuss/dss: docs/operations/cleanup.md

# Database cleanup

> With time, data will accumulate overtime in the database, leading to lot of old entries and lower performances.

Reviewer: ```suggestion
Data will accumulate over time in the database if client USSs do not remove their expired entities, and this can lead to lower performance due to quantity of stale entities.  For this reason, InterUSS recommends periodically cleaning up no-longer-relevant entities if USS clients do not always clean up after themselves.
```

This page describes how to clean up expired entities from the DSS datastore using the `db-manager evict` command.

## Overview

The `evict` subcommand of `db-manager` is a CLI tool that lists and deletes expired entities in the DSS store. The following entity types are supported:

- SCD operational intents
- SCD subscriptions
- RID identification service areas (ISAs)
- RID subscriptions

By default, the tool only lists expired entities. Deletion is opt-in via the `--delete` flag.

!!! warning
    Using this tool incorrectly may result in **loss of data**. Before deleting anything:

    - Always review and validate the list of entities identified as expired (run the tool without `--delete` first).
    - Ensure a backup of the data is available.
    - Double-check the TTL values passed to `--rid_ttl` and `--scd_ttl`.

> Expiration of entities is preferably determined through their end times. When end times are not available, the last update times are used instead.

Reviewer: I think end times should never not be available -- although the SCD schema seems to not have end_time as NOT NULL, it probably should be NOT NULL as [no unspecified end times are allowed](https://github.com/interuss/dss/blob/35bdeebdf56997bc3d69d17495a81720408956d1/pkg/scd/operational_intents_handler.go#L480)

```suggestion
Expiration of entities is preferably determined through their end times. In the unusual event that an end time is not available, the last update time is used instead.
```

Author: Yes, that coming from old documentation. Should we track it as an issue to have time not nullables?

Reviewer: Maybe as a P3.  I think the important thing here is noting that it would be unusual/unexpected.

Author: Done in https://github.com/interuss/dss/issues/1492

## Why and when to run the cleanup

> Expired entities (operational intents past their end time, stale subscriptions, ISAs whose lifetime has elapsed) accumulate over time in the datastore. While the DSS keeps functioning correctly with stale rows present, letting them pile up has real costs:

Reviewer: ```suggestion
Expired entities (operational intents past their end time, stale subscriptions, ISAs whose lifetime has elapsed) can accumulate over time in the datastore if not cleaned up by clients. While the DSS keeps functioning correctly with some stale rows present, excessive accumulation can impact:
```

> - Storage growth: unbounded table size on CockroachDB / Yugabytes.

Reviewer: ```suggestion
- Storage growth: unbounded storage usage in data stores (e.g., CockroachDB / YugabyteDB).
```

- Query performance: indexes get larger, range scans on `operational_intents` / `subscriptions` / `identification_service_areas` degrade.

> There is no single correct interval or TTL. Reasonable values depend on context and must be defined per DSS pool, taking into account: traffic volume and entity churn, regulatory or contractual data-retention requirements applicable to your jurisdiction, the storage capacity of the datastore cluster, and how long you may legitimately need to query historical entities for analysis or audit.

Reviewer: ```suggestion
There is no single correct interval or TTL. Reasonable values depend on context and must be defined per DSS pool, taking into account: traffic volume and entity churn, regulatory or contractual data-retention requirements applicable to your jurisdiction, the storage capacity of the datastore cluster, and how long clients may legitimately need to query historical entities.
```

> The defaults shipped with the deployment tooling (`30m` TTL on RID running every 30 min, `2688h` ≈ 56-day TTL on SCD running nightly when enabled) are starting points, not recommendations. Validate them against the criteria above and adjust then in production. When in doubt, start with conservative (longer) TTLs and tighten them as you gain confidence.

Reviewer: ```suggestion
The defaults shipped with the deployment tooling (`30m` TTL on RID running every 30 min, `2688h` ≈ 56-day TTL on SCD running nightly when enabled) are starting points, not recommendations. Validate production values against the criteria above.
```

## Performance impact

All expired entities are identified and removed within a single transaction. When the system is under heavy load, lock contention with concurrent transactions may cause the cleanup to fail. There is no risk of data inconsistency in this case - the cleanup may simply be retried.

To mitigate this:

- Run the cleanup during low-intensity periods (e.g. at night).
- Clean up iteratively, starting with a lower TTL and progressively increasing it.
If this becomes a recurring issue, batching removals could be considered as a future improvement.

## Usage

> Extract from `db-manager evict --help`:

Reviewer: ```suggestion
Extracted from `db-manager evict --help`:
```

      --datastore_max_conn_idle_secs int    maximum amount of time in seconds a connection may be idle, default is 30 seconds (default 30)
      --datastore_max_open_conns int        maximum number of open connections to the database, default is 4 (default 4)
      --datastore_max_retries int           maximum number of attempts to retry a query in case of contention, default is 100 (default 100)
      --datastore_port int                  database port to connect to (default 26257)
      --datastore_ssl_dir string            directory to ssl certificates. Must contain files: ca.crt, client.<user>.crt, client.<user>.key
      --datastore_ssl_mode string           database sslmode (default "disable")
      --datastore_user string               database user to authenticate as (default "root")
```

Notes:

- By default, expired entities are only listed - `--delete` is required to actually remove them.
- `--rid_ttl` and `--scd_ttl` accept durations formatted as [Go `time.Duration` strings](https://pkg.go.dev/time#ParseDuration), e.g. `24h`.
- The datastore connection flags match those of the `core-service` command.

## Regular cleanup

> Beyond running `db-manager evict` manually, the DSS deployment tooling can schedule the cleanup as a recurring Kubernetes `CronJob`. Three deployment paths expose the same set of evict knobs, but will always run with the  `--delete` flat set.

Reviewer: ```suggestion
Beyond running `db-manager evict` manually, the DSS deployment tooling can schedule the cleanup as a recurring Kubernetes `CronJob`. Three deployment paths expose the same set of evict knobs, but will always run with the  `--delete` flag set.
```

> Shared default: RID cleanup is enabled by default (`*/30 * * * *`, `ttl = 30m`); SCD cleanup is disabled by default (suggested schedule `0 2 * * *`, `ttl = 2688h` - i.e. 2 x 56 days).

Reviewer: @mickmis could you double-check that my suggestion is accurate?  I think the reason the defaults are the way they are is that each DSS instance is supposed to be responsible for cleaning up its own RID stuff (and deletion is limited to entities created by the DSS instance doing the deletion, as determined by locality), but SCD cleanup is a global activity and therefore DSS instance operators would want to at least coordinate to avoid cleaning up more than necessary (or perhaps designated a single USS as responsible for cleanup).

```suggestion
Shared default: RID cleanup is locality-limited and therefore is enabled by default (`*/30 * * * *`, `ttl = 30m`); SCD cleanup is not locality-limited and therefore disabled by default (suggested schedule `0 2 * * *`, `ttl = 2688h` - i.e. 2 x 56 days).
```

Reviewer 2: Your suggestion is correct, there is no `writer` column for SCD entities in order to store the locality, while there is for RID entities.

Do note that it is a possible scenario that some RID entities are inserted in the DB without any locality (either because of an older DSS version not supporting this, or because the locality not or mis-configured), in which case some entities might never get deleted. It could be worth a line here?

Author: The cleanup tool allow for an empty locality and will correctly clean up then no?

As configured, each DSS instance has it own cron, meaning old ones or one without a locality will just cleanup those without locality, so that probably fine no?

Reviewer 2: > The cleanup tool allow for an empty locality and will correctly clean up then no?

Yes, but you need to call it explicitly without any locality set (or an empty one). If there are any leftover entities without a locality but that all DSS instances are only cleaning up entities with their locality set, the leftover ones might remain indefinitely.

Author: Yes ok, but than that not different than a DSS changing locality: there will be leftover if no one run cleanup manually, that probably what we want to document no?

Author: I updated the documentation with:

- Rational of default
- Warning about locality changes

https://github.com/interuss/dss/compare/bd8a00f77e2c4aaef1737f89aa39e95209eda3e6..6cfa425115d56ba412c2279e3bf45ef1fdecd0e6

Reviewer 2: >  there will be leftover if no one run cleanup manually, that probably what we want to document no?

Yup

Reviewer 2: Maybe as well a note specific to empty locality, that can happen easily during a migration from no locality feature to locality is present

Author: Done

 Configure it under the `evict` key of your environment metadata:

```jsonnet
evict+: {
    scd+: {
        enable_cron: false,
        schedule: "0 2 * * *",
        ttl: "2688h",
        operational_intents: true,
        subscriptions: true,
    },
    rid+: {
        enable_cron: true,
        schedule: "*/30 * * * *",
        ttl: "30m",
        ISAs: true,
        subscriptions: true,
    },
},
```

### Terraform

When deploying via terrafrom modules, the parameters are configurable with module variables:

| Terraform variable              |  Default          |
|---------------------------------|------------------|
| `evict_enable_scd_cron`         | `false`          |
| `evict_scd_schedule`            | `"0 2 * * *"`    |
| `evict_scd_ttl`                 | `"2688h"`        |
| `evict_scd_operational_intents` | `true`           |
| `evict_scd_subscriptions`       | `true`           |
| `evict_enable_rid_cron`         | `true`           |
| `evict_rid_schedule`            | `"*/30 * * * *"` |
| `evict_rid_ttl`                 | `"30m"`          |
| `evict_rid_isas`                | `true`           |
| `evict_rid_subscriptions`       | `true`           |

## Examples

> The examples below assume a DSS running locally via the `run_locally.sh` script].

Reviewer: ```suggestion
The examples below assume a DSS running locally via the `run_locally.sh` script.
```

- Clean up iteratively, starting with a lower TTL and progressively increasing it.
If this becomes a recurring issue, batching removals could be considered as a future improvement.

> ## Changes in locally

Reviewer 2: ```suggestion
## Changes in locality
```

> There may be cases where a DSS instance changes its locality. A common scenario is that locality was not previously mandatory, though regular updates may also occur.

Author: ```suggestion
There may be cases where a DSS instance changes its locality. This commonly occurs if locality wasn't previously required, though routine updates can also trigger a change.
```
